File: inspector/frontend/src/tabs/segments/domain/registry.ts
/**
 * Issue Registry — TypeScript twin of
 * ``inspector/services/validation/registry.py``.
 *
 * Each row is the single source of truth for one validation category's
 * UI / persistence / suppression metadata. The Python and TS sides MUST
 * agree row-for-row; ``__tests__/registry/parity.test.ts`` enforces this.
 *
 * camelCase mirrors of the Python fields:
 *   kind / cardType / severity / accordionOrder / canIgnore /
 *   autoSuppress / persistsIgnore / scope / displayTitle / description.
 */

import type { SortOption } from './sorting';

export type CardType = 'generic' | 'missingWords' | 'missingVerses' | 'error';
export type Severity = 'error' | 'warning' | 'info';
export type Scope = 'per_segment' | 'per_verse' | 'per_chapter';

export interface IssueDefinition {
    kind: string;
    cardType: CardType;
    severity: Severity;
    accordionOrder: number;
    canIgnore: boolean;
    autoSuppress: boolean;
    persistsIgnore: boolean;
    scope: Scope;
    displayTitle: string;
    description: string;
    /**
     * Sort options the accordion offers (first = active default). FE-only
     * presentation concern — deliberately absent from the Python registry and
     * its parity snapshot. Omit to offer no sorting (Missing Verses). See
     * `./sorting.ts`.
     */
    sorts?: readonly SortOption[];
}

export const IssueRegistry: Readonly<Record<string, IssueDefinition>> = Object.freeze({
    failed: {
        kind: 'failed',
        cardType: 'generic',
        severity: 'error',
        accordionOrder: 1,
        canIgnore: false,
        autoSuppress: true,
        persistsIgnore: false,
        scope: 'per_segment',
        displayTitle: 'Failed Alignments',
        description: '',
        sorts: [{ kind: 'quran_order', default: true }],
    },
    missing_verses: {
        kind: 'missing_verses',
        cardType: 'missingVerses',
        severity: 'error',
        accordionOrder: 2,
        canIgnore: false,
        autoSuppress: true,
        persistsIgnore: false,
        scope: 'per_verse',
        displayTitle: 'Missing Verses',
        description: '',
    },
    missing_words: {
        kind: 'missing_words',
        cardType: 'missingWords',
        severity: 'error',
        accordionOrder: 3,
        canIgnore: false,
        autoSuppress: false,
        persistsIgnore: false,
        scope: 'per_verse',
        displayTitle: 'Missing Words',
        description: '',
        sorts: [{ kind: 'quran_order', default: true }, { kind: 'word_count' }],
    },
    structural_errors: {
        kind: 'structural_errors',
        cardType: 'error',
        severity: 'error',
        accordionOrder: 4,
        canIgnore: false,
        autoSuppress: true,
        persistsIgnore: false,
        scope: 'per_chapter',
        displayTitle: 'Structural Errors',
        description: '',
        sorts: [{ kind: 'quran_order', default: true }],
    },
    low_confidence: {
        kind: 'low_confidence',
        cardType: 'generic',
        severity: 'warning',
        accordionOrder: 5,
        canIgnore: true,
        autoSuppress: true,
        persistsIgnore: true,
        scope: 'per_segment',
        displayTitle: 'Low Confidence',
        description: '',
        sorts: [{ kind: 'confidence', default: true }, { kind: 'quran_order' }],
    },
    low_confidence_v2: {
        kind: 'low_confidence_v2',
        cardType: 'generic',
        severity: 'warning',
        accordionOrder: 6,
        canIgnore: true,
        autoSuppress: true,
        persistsIgnore: true,
        scope: 'per_segment',
        displayTitle: 'Low Confidence v2',
        description: 'MFA tight-beam probe disagreed with the DP alignment for these segments. Treat as a second-opinion warning.',
        sorts: [{ kind: 'quran_order', default: true }],
    },
    repetitions: {
        kind: 'repetitions',
        cardType: 'generic',
        severity: 'warning',
        accordionOrder: 9,
        canIgnore: true,
        autoSuppress: true,
        persistsIgnore: true,
        scope: 'per_segment',
        displayTitle: 'Detected Repetitions',
        description: '',
        sorts: [{ kind: 'quran_order', default: true }, { kind: 'rep_split_count' }],
    },
    audio_bleeding: {
        kind: 'audio_bleeding',
        cardType: 'generic',
        severity: 'warning',
        accordionOrder: 7,
        canIgnore: true,
        autoSuppress: true,
        persistsIgnore: true,
        scope: 'per_segment',
        displayTitle: 'Audio Bleeding',
        description: '',
        sorts: [{ kind: 'quran_order', default: true }],
    },
    boundary_adj: {
        kind: 'boundary_adj',
        cardType: 'generic',
        severity: 'warning',
        accordionOrder: 8,
        canIgnore: true,
        autoSuppress: true,
        persistsIgnore: true,
        scope: 'per_segment',
        displayTitle: 'May Require Boundary Adjustment',
        description: '',
        sorts: [{ kind: 'quran_order', default: true }],
    },
    cross_verse: {
        kind: 'cross_verse',
        cardType: 'generic',
        severity: 'warning',
        accordionOrder: 10,
        canIgnore: false,
        autoSuppress: false,
        persistsIgnore: false,
        scope: 'per_segment',
        displayTitle: 'Cross-verse',
        description: 'Split cross verses at the verse boundary. If the reciter recited continuously across the boundary, annotate the new inter-segment boundary as wasl.',
        sorts: [{ kind: 'quran_order', default: true }, { kind: 'verse_count' }],
    },
    qalqala: {
        kind: 'qalqala',
        cardType: 'generic',
        severity: 'info',
        accordionOrder: 11,
        canIgnore: false,
        autoSuppress: false,
        persistsIgnore: false,
        scope: 'per_segment',
        displayTitle: 'Qalqala',
        description: '',
        sorts: [{ kind: 'quran_order', default: true }],
    },
    muqattaat: {
        kind: 'muqattaat',
        cardType: 'generic',
        severity: 'info',
        accordionOrder: 12,
        canIgnore: false,
        autoSuppress: false,
        persistsIgnore: false,
        scope: 'per_segment',
        displayTitle: 'Muqattaʼat',
        description: '',
        sorts: [{ kind: 'quran_order', default: true }],
    },
    basmala_amin: {
        kind: 'basmala_amin',
        cardType: 'generic',
        severity: 'info',
        accordionOrder: 13,
        canIgnore: true,
        autoSuppress: true,
        persistsIgnore: true,
        scope: 'per_segment',
        displayTitle: 'Basmala + Amin',
        description: '',
        sorts: [{ kind: 'quran_order', default: true }],
    },
    hidden_pause: {
        kind: 'hidden_pause',
        cardType: 'generic',
        severity: 'info',
        accordionOrder: 14,
        canIgnore: true,
        autoSuppress: true,
        persistsIgnore: true,
        scope: 'per_segment',
        displayTitle: 'Hidden Pause (review)',
        description: 'Re-segmentation found a pause inside this segment. Auto Split places the proposed cut; ignore if there is no pause.',
        sorts: [{ kind: 'score', default: true }, { kind: 'quran_order' }],
    },
    false_split: {
        kind: 'false_split',
        cardType: 'generic',
        severity: 'info',
        accordionOrder: 15,
        canIgnore: true,
        autoSuppress: true,
        persistsIgnore: true,
        scope: 'per_segment',
        displayTitle: 'False Split (review)',
        description: "Re-segmentation heard continuous speech across this segment's end. Merge with the next segment if the cut is inside speech; ignore if the pause is real.",
        sorts: [{ kind: 'score', default: true }, { kind: 'quran_order' }],
    },
});

const _entries = Object.entries(IssueRegistry) as [string, IssueDefinition][];

export const ALL_CATEGORIES: readonly string[] = _entries.map(([k]) => k);
export const PER_SEGMENT_CATEGORIES: readonly string[] = _entries
    .filter(([, v]) => v.scope === 'per_segment').map(([k]) => k);
export const PER_VERSE_CATEGORIES: readonly string[] = _entries
    .filter(([, v]) => v.scope === 'per_verse').map(([k]) => k);
export const PER_CHAPTER_CATEGORIES: readonly string[] = _entries
    .filter(([, v]) => v.scope === 'per_chapter').map(([k]) => k);
export const CAN_IGNORE_CATEGORIES: readonly string[] = _entries
    .filter(([, v]) => v.canIgnore).map(([k]) => k);
export const AUTO_SUPPRESS_CATEGORIES: readonly string[] = _entries
    .filter(([, v]) => v.autoSuppress).map(([k]) => k);
export const PERSISTS_IGNORE_CATEGORIES: readonly string[] = _entries
    .filter(([, v]) => v.persistsIgnore).map(([k]) => k);

/**
 * Drop categories whose registry entry has ``persistsIgnore=false``.
 * The legacy ``"_all"`` marker passes through unchanged.
 */
export function filterPersistentIgnores(categories: readonly string[] | undefined | null): string[] {
    if (!categories) return [];
    const out: string[] = [];
    for (const cat of categories) {
        if (cat === '_all') {
            out.push(cat);
            continue;
        }
        const defn = IssueRegistry[cat];
        if (!defn || defn.persistsIgnore) out.push(cat);
    }
    return out;
}

/** Display titles indexed by category — derived for UI labels. */
export const ERROR_CAT_LABELS: Readonly<Record<string, string>> = Object.freeze(
    Object.fromEntries(_entries.map(([k, v]) => [k, v.displayTitle])),
);

File: inspector/services/validation/registry.py
"""Issue Registry: single source of truth for validation-category metadata.

Each row pins:
- ``kind``               — registry key (matches the dict key).
- ``card_type``          — UI dispatch tag for the validation card subcomponent.
- ``severity``           — ``"error" | "warning" | "info"``.
- ``accordion_order``    — 1-based render order in the validation accordion.
- ``can_ignore``         — base gate for the Ignore button on the category card.
- ``auto_suppress``      — whether an edit launched from this card writes the
                           category into ``seg.ignored_categories`` (per-segment
                           scope only; chapter / verse scopes treat this flag as
                           declarative, since revalidation is the source of
                           truth there).
- ``persists_ignore``    — whether the category survives save serialization.
                           ``False`` means the category is filtered out of
                           ``ignored_categories`` before persisting to
                           ``detailed.json``.
- ``scope``              — granularity of the issue: ``"per_segment"``,
                           ``"per_verse"``, or ``"per_chapter"``.
- ``display_title``      — user-facing accordion header for the category.
- ``description``        — short sentence shown in tooltips / help; may be
                           empty when the title is self-explanatory.

The registry is also exposed as a hand-mirrored TypeScript twin at
``inspector/frontend/src/tabs/segments/domain/registry.ts``. The two sides are
asserted equal by ``__tests__/registry/parity.test.ts``.
"""

from __future__ import annotations

from collections.abc import Iterator, Mapping
from dataclasses import asdict, dataclass
from typing import Any, Literal

CardType = Literal["generic", "missingWords", "missingVerses", "error"]
Severity = Literal["error", "warning", "info"]
Scope = Literal["per_segment", "per_verse", "per_chapter"]


@dataclass(frozen=True)
class IssueDefinition:
    """Frozen row describing one validation category."""

    kind: str
    card_type: CardType
    severity: Severity
    accordion_order: int
    can_ignore: bool
    auto_suppress: bool
    persists_ignore: bool
    scope: Scope
    display_title: str
    description: str

    # Permit ``row["field"]`` access alongside ``row.field`` so registry rows
    # interoperate with monkeypatched plain-dict rows used in the extensibility
    # test (``test_registry_extensibility.py``) and with future JSON snapshots.
    def __getitem__(self, key: str) -> Any:
        return getattr(self, key)


_REGISTRY: dict[str, IssueDefinition] = {
    "failed": IssueDefinition(
        kind="failed",
        card_type="generic",
        severity="error",
        accordion_order=1,
        can_ignore=False,
        auto_suppress=True,
        persists_ignore=False,
        scope="per_segment",
        display_title="Failed Alignments",
        description="",
    ),
    "missing_verses": IssueDefinition(
        kind="missing_verses",
        card_type="missingVerses",
        severity="error",
        accordion_order=2,
        can_ignore=False,
        auto_suppress=True,
        persists_ignore=False,
        scope="per_verse",
        display_title="Missing Verses",
        description="",
    ),
    "missing_words": IssueDefinition(
        kind="missing_words",
        card_type="missingWords",
        severity="error",
        accordion_order=3,
        can_ignore=False,
        auto_suppress=False,
        persists_ignore=False,
        scope="per_verse",
        display_title="Missing Words",
        description="",
    ),
    "structural_errors": IssueDefinition(
        kind="structural_errors",
        card_type="error",
        severity="error",
        accordion_order=4,
        can_ignore=False,
        auto_suppress=True,
        persists_ignore=False,
        scope="per_chapter",
        display_title="Structural Errors",
        description="",
    ),
    "low_confidence": IssueDefinition(
        kind="low_confidence",
        card_type="generic",
        severity="warning",
        accordion_order=5,
        can_ignore=True,
        auto_suppress=True,
        persists_ignore=True,
        scope="per_segment",
        display_title="Low Confidence",
        description="",
    ),
    "low_confidence_v2": IssueDefinition(
        kind="low_confidence_v2",
        card_type="generic",
        severity="warning",
        accordion_order=6,
        can_ignore=True,
        auto_suppress=True,
        persists_ignore=True,
        scope="per_segment",
        display_title="Low Confidence v2",
        description="MFA tight-beam probe disagreed with the DP alignment for these segments. Treat as a second-opinion warning.",
    ),
    "repetitions": IssueDefinition(
        kind="repetitions",
        card_type="generic",
        severity="warning",
        accordion_order=9,
        can_ignore=True,
        auto_suppress=True,
        persists_ignore=True,
        scope="per_segment",
        display_title="Detected Repetitions",
        description="Detected part of the text repeated in the audio. Split if so, ignore if a false flag.",
    ),
    "audio_bleeding": IssueDefinition(
        kind="audio_bleeding",
        card_type="generic",
        severity="warning",
        accordion_order=7,
        can_ignore=True,
        auto_suppress=True,
        persists_ignore=True,
        scope="per_segment",
        display_title="Audio Bleeding",
        description="",
    ),
    "boundary_adj": IssueDefinition(
        kind="boundary_adj",
        card_type="generic",
        severity="warning",
        accordion_order=8,
        can_ignore=True,
        auto_suppress=True,
        persists_ignore=True,
        scope="per_segment",
        display_title="May Require Boundary Adjustment",
        description="",
    ),
    "cross_verse": IssueDefinition(
        kind="cross_verse",
        card_type="generic",
        severity="warning",
        accordion_order=10,
        can_ignore=False,
        auto_suppress=False,
        persists_ignore=False,
        scope="per_segment",
        display_title="Cross-verse",
        description="Split cross verses at the verse boundary. If the reciter recited continuously across the boundary, annotate the new inter-segment boundary as wasl. Auto split is for speeding things up.",
    ),
    "qalqala": IssueDefinition(
        kind="qalqala",
        card_type="generic",
        severity="info",
        accordion_order=11,
        can_ignore=False,
        auto_suppress=False,
        persists_ignore=False,
        scope="per_segment",
        display_title="Qalqala",
        description="",
    ),
    "muqattaat": IssueDefinition(
        kind="muqattaat",
        card_type="generic",
        severity="info",
        accordion_order=12,
        can_ignore=False,
        auto_suppress=False,
        persists_ignore=False,
        scope="per_segment",
        display_title="Muqattaʼat",
        description="",
    ),
    "basmala_amin": IssueDefinition(
        kind="basmala_amin",
        card_type="generic",
        severity="info",
        accordion_order=13,
        can_ignore=True,
        auto_suppress=True,
        persists_ignore=True,
        scope="per_segment",
        display_title="Basmala + Amin",
        description="",
    ),
    "hidden_pause": IssueDefinition(
        kind="hidden_pause",
        card_type="generic",
        severity="info",
        accordion_order=14,
        can_ignore=True,
        auto_suppress=True,
        persists_ignore=True,
        scope="per_segment",
        display_title="Hidden Pause (review)",
        description="Re-segmentation found a pause inside this segment. Auto Split places the proposed cut; ignore if there is no pause.",
    ),
    "false_split": IssueDefinition(
        kind="false_split",
        card_type="generic",
        severity="info",
        accordion_order=15,
        can_ignore=True,
        auto_suppress=True,
        persists_ignore=True,
        scope="per_segment",
        display_title="False Split (review)",
        description="Re-segmentation heard continuous speech across this segment's end. Merge with the next segment if the cut is inside speech; ignore if the pause is real.",
    ),
}


class _RegistryView(Mapping[str, IssueDefinition]):
    """Mapping facade over ``_REGISTRY`` exposing a stable public API.

    Tests rely on ``IssueRegistry[cat]``, ``IssueRegistry.keys()``, and the
    ``_registry`` attribute (used by ``test_registry_extensibility.py`` to
    monkeypatch a synthetic row in).
    """

    def __init__(self, store: dict[str, IssueDefinition]) -> None:
        self._registry = store

    def __getitem__(self, key: str) -> IssueDefinition:
        return self._registry[key]

    def __iter__(self) -> Iterator[str]:
        return iter(self._registry)

    def __len__(self) -> int:
        return len(self._registry)

    def __contains__(self, key: object) -> bool:
        return key in self._registry


IssueRegistry: _RegistryView = _RegistryView(_REGISTRY)


ALL_CATEGORIES: tuple[str, ...] = tuple(_REGISTRY.keys())
PER_SEGMENT_CATEGORIES: tuple[str, ...] = tuple(
    k for k, v in _REGISTRY.items() if v.scope == "per_segment"
)
PER_VERSE_CATEGORIES: tuple[str, ...] = tuple(
    k for k, v in _REGISTRY.items() if v.scope == "per_verse"
)
PER_CHAPTER_CATEGORIES: tuple[str, ...] = tuple(
    k for k, v in _REGISTRY.items() if v.scope == "per_chapter"
)
CAN_IGNORE_CATEGORIES: tuple[str, ...] = tuple(k for k, v in _REGISTRY.items() if v.can_ignore)
AUTO_SUPPRESS_CATEGORIES: tuple[str, ...] = tuple(
    k for k, v in _REGISTRY.items() if v.auto_suppress
)
PERSISTS_IGNORE_CATEGORIES: tuple[str, ...] = tuple(
    k for k, v in _REGISTRY.items() if v.persists_ignore
)


def filter_persistent_ignores(categories: list[str] | None) -> list[str]:
    """Return ``categories`` minus any whose registry entry has
    ``persists_ignore=False``.

    The legacy ``"_all"`` marker passes through unchanged — it predates the
    per-category registry and represents a session-level "ignore everything"
    flag.
    """
    if not categories:
        return []
    out: list[str] = []
    for cat in categories:
        if cat == "_all":
            out.append(cat)
            continue
        defn = _REGISTRY.get(cat)
        if defn is None or defn.persists_ignore:
            out.append(cat)
    return out


def registry_as_dict() -> dict[str, dict[str, Any]]:
    """Serialize the registry to a plain dict (for parity snapshots / fixtures)."""
    return {k: asdict(v) for k, v in _REGISTRY.items()}


__all__ = [
    "IssueDefinition",
    "IssueRegistry",
    "ALL_CATEGORIES",
    "PER_SEGMENT_CATEGORIES",
    "PER_VERSE_CATEGORIES",
    "PER_CHAPTER_CATEGORIES",
    "CAN_IGNORE_CATEGORIES",
    "AUTO_SUPPRESS_CATEGORIES",
    "PERSISTS_IGNORE_CATEGORIES",
    "filter_persistent_ignores",
    "registry_as_dict",
]
